packages/runtime/datastore/src/dataStoreRuntime.ts

```diff
@@ -377,38 +391,52 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
         if (this.isAttached) {
             this.attachChannel(channel);
             return;
-        } else {
-            this.bind(channel.handle);
+        }
 
-            // If our data store is local then add the channel to the queue
-            if (!this.localChannelContextQueue.has(channel.id)) {
-                this.localChannelContextQueue.set(channel.id, this.contexts.get(channel.id) as LocalChannelContextBase);
-            }
+        /**
+         * If this channel has already been bound, do nothing. This can happen during attachGraph() when a channel's
+         * graph is attached. It calls bindToContext on the shared object which will end up back here.
+         */
+        if (this.boundhandles !== undefined && this.boundhandles.has(channel.handle)) {
+            return;
+        }
+
+        this.bind(channel.handle);
+
+        // If our data store is local then add the channel to the queue
+        if (!this.localChannelContextQueue.has(channel.id)) {
+            this.localChannelContextQueue.set(channel.id, this.contexts.get(channel.id) as LocalChannelContextBase);
         }
     }
 
-    public attachGraph() {
-        if (this.graphAttachState !== AttachState.Detached) {
+    /**
+     * This function is called when a data store becomes root. It does the following:
+     * 1. Marks the data store locally visible in the container.
+     * 2. Attaches the graph of all the handles bound to it.
+     * 3. Calls into the data store context to mark it visible in the container too. If the container is globally
+     *    visible, it will mark us globally visible. Otherwise, it will mark us globally visible when it becomes
+     *    globally visible.
+     */
+    public makeVisibleAndAttachGraph() {
+        if (this.visibilityState !== VisibilityState.NotVisible) {
```

Comment:
no, data store runtime only cares about making itself locally visible. It will become globally visible if the container is attached and this is decided by the container runtime.